v-next/hardhat/src/internal/builtin-plugins/solidity-test/task-action.ts

```diff
@@ -19,25 +31,46 @@ const runSolidityTests: NewTaskActionFunction<TestActionArguments> = async (
   { timeout, noCompile },
   hre,
 ) => {
-  if (!noCompile) {
-    await hre.tasks.getTask("compile").run({});
-    console.log();
-  }
-
-  const artifacts = await getArtifacts(hre.artifacts);
-  const testSuiteIds = (
+  const rootFilePaths = (
     await Promise.all(
-      artifacts.map(async (artifact) => {
-        if (await isTestArtifact(hre.config.paths.root, artifact)) {
-          return artifact.id;
-        }
-      }),
+      hre.config.paths.tests.solidity
+        .map((p) => resolveFromRoot(hre.config.paths.root, p))
+        .map(async (p) => {
+          // NOTE: The paths specified in the `paths.tests.solidity` array
+          // can be either directories or files.
+          if (await isDirectory(p)) {
+            return getAllFilesMatching(p, (f) => f.endsWith(".sol"));
+          } else if (p.endsWith(".sol") === true) {
+            return [p];
+          } else {
+            return [];
+          }
+        }),
     )
-  ).filter((artifact) => artifact !== undefined);
+  ).flat(1);
+
+  const buildOptions: BuildOptions = {
+    // NOTE: The uncached sources will still be compiled event if `noCompile`
+    // is true. We could consider adding a `cacheOnly` option to support true
+    // `noCompile` behavior.
+    force: !noCompile,
```

Comment:
If I run `npx hardhat test solidity` it will force and recompile everything

Reply:
Maybe add an issue and address it after we have the compilation cache in place, and when we have the split between production and test compilation caches or while implementing that.

Reply:
We should also consider removing this option